### PR TITLE
Meta module tweaks

### PR DESCRIPTION
**Pull request by natebosch**
- Avoid `(toList()..sort()).first`, use `reduce(_min)` instead.
- Add a `merge` method to `Module` so it can be used from multiple
  places.
- Don't make copies of collections that are unreferenced later.
- Use `Iterable<Module>` where possible instead of `Set<AssetId>` or
  `Map<AssetId, Module>`
- Keep separate collections for the different varieties of modules and
  concatenate them after merging.
- Change some for loops to `map` when we care more about the collection
  than the side effects.

### CODE REVIEW

File: build_modules/lib/src/meta_module.dart [hunk 1]
```diff
@@ -167,68 +151,86 @@ Set<AssetId> _localTransitiveDeps(
...
 ///
-///   * If it is an entrypoint module, skip it.
-///   * Else merge it into a module whose name is a the entrypoint
-///   that import it (create that module if it doesn't exist).
-List<Module> _mergeModules(
-    Map<AssetId, Module> originalModulesById, Set<AssetId> entrypoints) {
-  var modulesById = new Map<AssetId, Module>.from(originalModulesById);
+///   * If it is an entrypoint module do not merge it.
+///   * If it is not depended on my any entrypoint do not merge it.
+///   * If it is depended on by on entrypoint merge it into the entrypoint
```
Comment (grouma) [inline build_modules/lib/src/meta_module.dart]: typo

File: build_modules/lib/src/meta_module.dart [hunk 2]
```diff
@@ -167,68 +151,86 @@ Set<AssetId> _localTransitiveDeps(
...
+///     modules
+///   * Else merge it into with others that are depended on by the same set of
+///     entrypoints
```
Comment (grouma) [inline build_modules/lib/src/meta_module.dart]: Dart highlighters act funny if you indent past the bullet point. Keep this under the bullet.

File: build_modules/lib/src/meta_module.dart (hunk 1 above)
Comment (natebosch) [reply build_modules/lib/src/meta_module.dart]: Done

File: build_modules/lib/src/meta_module.dart (hunk 2 above)
Comment (natebosch) [reply build_modules/lib/src/meta_module.dart]: Done

**Comment by natebosch** [comment]
Pushing on red because the tests are flaky but this commit is not introducing the flakes and this makes the code easier to understand while we are debugging.

**Pull request merged**